### PR TITLE
Build: Enhance build system

### PR DESCRIPTION
**Pull request by pan3793**
# [INFRA] Enhance build system

### What changes were proposed in this pull request?

- `build/mvn` is introduced for auto-downloading Apache Maven if no suitable candidate `mvn` in the PATH, the suitable `mvn` means the version of `mvn` must be equal or great than `maven.version` defined in `pom.xml`
- `spark-2` and `spark-3` are changed into `spark-2.4` and `spark-3.0`, another versions like `spark-3.1`, `spark-3.2` will be added in the following PRs.
- simplify GitHub workflow yaml by using matrix syntax
- enable build cache on GitHub workflow
- extract version of most deps into maven properties

### Why are the changes needed?

To make `pom.xml` and CI script more clear and flexible to build w/ different spark versions

### How was this patch tested?

/cc @related-reviewer

/assign @main-reviewer

### CODE REVIEW

File: client-spark/shuffle-manager-2/src/main/scala/org/apache/spark/shuffle/rss/RssShuffleReader.scala [hunk 1]
```diff
@@ -110,7 +110,7 @@ class RssShuffleReader[K, C](
         context.taskMetrics().incDiskBytesSpilled(sorter.diskBytesSpilled)
         context.taskMetrics().incPeakExecutionMemory(sorter.peakMemoryUsedBytes)
         // Use completion callback to stop sorter if task was finished/cancelled.
-        context.addTaskCompletionListener[Unit](_ => {
+        context.addTaskCompletionListener(_ => {
```
Comment (pan3793) [inline client-spark/shuffle-manager-2/src/main/scala/org/apache/spark/shuffle/rss/RssShuffleReader.scala:113]: be compatible w/ spark 2.3

File: server-worker/src/test/scala/com/aliyun/emr/rss/service/deploy/integration/SkewJoinTest.scala [hunk 2]
```diff
@@ -29,10 +29,11 @@ class SkewJoinTest extends SparkTestBase {
       .set("spark.serializer", "org.apache.spark.serializer.KryoSerializer")
       .set("spark.sql.adaptive.autoBroadcastJoinThreshold", "-1")
       .set("spark.sql.autoBroadcastJoinThreshold", "-1")
+      .set("spark.sql.parquet.compression.codec", "gzip")
 
     enableRss(sparkConf)
 
-    val sparkSession = SparkSession.builder().enableHiveSupport().config(sparkConf).getOrCreate()
+    val sparkSession = SparkSession.builder().config(sparkConf).getOrCreate()
```
Comment (pan3793) [inline server-worker/src/test/scala/com/aliyun/emr/rss/service/deploy/integration/SkewJoinTest.scala:36]: the spark data source table is sufficient for this test case

File: server-worker/src/test/scala/com/aliyun/emr/rss/service/deploy/integration/SkewJoinTest.scala [hunk 3]
```diff
@@ -29,10 +29,11 @@ class SkewJoinTest extends SparkTestBase {
       .set("spark.serializer", "org.apache.spark.serializer.KryoSerializer")
       .set("spark.sql.adaptive.autoBroadcastJoinThreshold", "-1")
       .set("spark.sql.autoBroadcastJoinThreshold", "-1")
+      .set("spark.sql.parquet.compression.codec", "gzip")
```
Comment (pan3793) [inline server-worker/src/test/scala/com/aliyun/emr/rss/service/deploy/integration/SkewJoinTest.scala:32]: spark parquet table uses snappy in default, not sure why snappy dependency is broken here, just switching to gzip as a workaround.

File: build/mvn [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+      exit 2
+    cd "${_DIR}" && tar -xzf "$2"
+    rm -rf "$local_tarball"
+  fi
+}
+
+# Determine the Maven version from the root pom.xml file and
+# install maven under the build/ folder if needed.
+install_mvn() {
+  local MVN_VERSION=`grep "<maven.version>" "${_DIR}/../pom.xml" | head -n1 | awk -F '[<>]' '{print $3}'`
```
Comment (pan3793) [inline build/mvn:72]: keep mvn version consistent w/ pom.xml

File: client-spark/shuffle-manager-2/src/main/scala/org/apache/spark/shuffle/rss/RssShuffleReader.scala (hunk 1 above)
Comment (pan3793) [inline client-spark/shuffle-manager-2/src/main/scala/org/apache/spark/shuffle/rss/RssShuffleReader.scala:113]: Fix compile errpr w/ Spark 2.3.

File: pom.xml [hunk 5]
```diff
@@ -31,27 +31,48 @@
...
+    <grpc.version>1.44.0</grpc.version>
+    <guava.version>14.0.1</guava.version>
     <javaxservlet.version>3.1.0</javaxservlet.version>
-    <!-- Apache Ratis version -->
+    <junit.version>4.12</junit.version>
+    <leveldb.version>1.8</leveldb.version>
+    <log4j.version>1.2.17</log4j.version>
+    <log4j2.version>2.17.1</log4j2.version>
+    <lz4-java.version>1.8.0</lz4-java.version>
+    <mockito.version>1.10.19</mockito.version>
```
Comment (pan3793) [inline pom.xml:54]: I found different and incompatible mockito versions are used across modules, will try to unify them in the following PRs

**Comment by FMX** [comment]
LGTM.

**Comment by FMX** [comment]
LGTM. Thanks.

**Review by waitinfuture — APPROVED** [review]
LGTM, thanks for the nice work!

**Pull request merged**